File: backend/src/main/java/com/seuprojeto/tmdb/TmdbService.java
package com.seuprojeto.tmdb;

import com.fasterxml.jackson.databind.JsonNode;
import com.seuprojeto.catalog.*;
import com.seuprojeto.tmdb.dto.TmdbImportResult;
import com.seuprojeto.tmdb.dto.TmdbSearchResult;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import org.springframework.web.client.RestTemplate;
import org.springframework.web.util.UriComponentsBuilder;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

@Service
@RequiredArgsConstructor
public class TmdbService {

    private final RestTemplate tmdbRestTemplate;
    private final TmdbConfig.TmdbProperties tmdbProperties;
    private final ContentRepository contentRepository;
    private final SeasonRepository seasonRepository;
    private final EpisodeRepository episodeRepository;

    public boolean isConfigured() {
        return tmdbProperties.isConfigured();
    }

    public List<TmdbSearchResult> buscar(String query, ContentType tipo) {
        if (!isConfigured()) {
            throw new IllegalStateException(
                    "Integração com o TMDB não configurada. Defina a variável de ambiente TMDB_API_KEY.");
        }

        String path = tipo == ContentType.FILME ? "/search/movie" : "/search/tv";
        String url = UriComponentsBuilder.fromHttpUrl(tmdbProperties.getBaseUrl() + path)
                .queryParam("api_key", tmdbProperties.getApiKey())
                .queryParam("language", "pt-BR")
                .queryParam("query", query)
                .queryParam("include_adult", false)
                .build()
                .toUriString();

        JsonNode response = tmdbRestTemplate.getForObject(url, JsonNode.class);
        if (response == null || !response.has("results")) {
            return List.of();
        }

        return StreamSupport.stream(response.get("results").spliterator(), false)
                .limit(20)
                .map(node -> toSearchResult(node, tipo))
                .collect(Collectors.toList());
    }

    private TmdbSearchResult toSearchResult(JsonNode node, ContentType tipo) {
        String titulo = tipo == ContentType.FILME
                ? textOrNull(node, "title")
                : textOrNull(node, "name");

        String dataLancamento = tipo == ContentType.FILME
                ? textOrNull(node, "release_date")
                : textOrNull(node, "first_air_date");

        Integer ano = null;
        if (dataLancamento != null && dataLancamento.length() >= 4) {
            try {
                ano = Integer.parseInt(dataLancamento.substring(0, 4));
            } catch (NumberFormatException ignored) {
                // mantém ano nulo se a data vier em formato inesperado
            }
        }

        String posterPath = textOrNull(node, "poster_path");
        String imagemUrl = posterPath != null ? tmdbProperties.getImageBaseUrl() + posterPath : null;

        return TmdbSearchResult.builder()
                .tmdbId(node.get("id").asLong())
                .titulo(titulo != null ? titulo : "Sem título")
                .ano(ano)
                .imagemUrl(imagemUrl)
                .sinopse(textOrNull(node, "overview"))
                .tipo(tipo)
                .avaliacao(node.has("vote_average") ? node.get("vote_average").asDouble() : null)
                .build();
    }

    @Transactional
    public TmdbImportResult importar(Long tmdbId, ContentType tipo) {
        if (!isConfigured()) {
            throw new IllegalStateException(
                    "Integração com o TMDB não configurada. Defina a variável de ambiente TMDB_API_KEY.");
        }

        String path = tipo == ContentType.FILME ? "/movie/" + tmdbId : "/tv/" + tmdbId;
        String detailsUrl = UriComponentsBuilder.fromHttpUrl(tmdbProperties.getBaseUrl() + path)
                .queryParam("api_key", tmdbProperties.getApiKey())
                .queryParam("language", "pt-BR")
                .build()
                .toUriString();

        JsonNode details = tmdbRestTemplate.getForObject(detailsUrl, JsonNode.class);
        if (details == null) {
            throw new IllegalArgumentException("Não foi possível encontrar este título no TMDB");
        }

        String titulo = tipo == ContentType.FILME ? textOrNull(details, "title") : textOrNull(details, "name");
        String dataLancamento = tipo == ContentType.FILME
                ? textOrNull(details, "release_date")
                : textOrNull(details, "first_air_date");
        Integer ano = null;
        if (dataLancamento != null && dataLancamento.length() >= 4) {
            try {
                ano = Integer.parseInt(dataLancamento.substring(0, 4));
            } catch (NumberFormatException ignored) {
            }
        }

        String genero = null;
        if (details.has("genres") && details.get("genres").isArray()) {
            List<String> generos = new ArrayList<>();
            details.get("genres").forEach(g -> generos.add(g.get("name").asText()));
            genero = String.join(", ", generos);
        }

        String posterPath = textOrNull(details, "poster_path");
        String imagemUrl = posterPath != null ? tmdbProperties.getImageBaseUrl() + posterPath : null;

        Content content = Content.builder()
                .titulo(titulo != null ? titulo : "Sem título")
                .sinopse(textOrNull(details, "overview"))
                .genero(genero)
                .ano(ano)
                .imagemUrl(imagemUrl)
                .tipo(tipo)
                .assistido(false)
                .build();
        content = contentRepository.save(content);

        int temporadasImportadas = 0;
        int episodiosImportados = 0;

        if (tipo == ContentType.SERIE && details.has("seasons")) {
            for (JsonNode seasonNode : details.get("seasons")) {
                int seasonNumber = seasonNode.get("season_number").asInt();

                Season season = Season.builder()
                        .numero(seasonNumber)
                        .titulo(textOrNull(seasonNode, "name"))
                        .conteudo(content)
                        .build();
                season = seasonRepository.save(season);
                temporadasImportadas++;

                episodiosImportados += importarEpisodios(tmdbId, seasonNumber, season);
            }
        }

        return TmdbImportResult.builder()
                .conteudoId(content.getId())
                .titulo(content.getTitulo())
                .temporadasImportadas(temporadasImportadas)
                .episodiosImportados(episodiosImportados)
                .build();
    }

    private int importarEpisodios(Long tmdbId, int seasonNumber, Season season) {
        String episodesUrl = UriComponentsBuilder
                .fromHttpUrl(tmdbProperties.getBaseUrl() + "/tv/" + tmdbId + "/season/" + seasonNumber)
                .queryParam("api_key", tmdbProperties.getApiKey())
                .queryParam("language", "pt-BR")
                .build()
                .toUriString();

        JsonNode seasonDetails;
        try {
            seasonDetails = tmdbRestTemplate.getForObject(episodesUrl, JsonNode.class);
        } catch (Exception e) {
            return 0;
        }

        if (seasonDetails == null || !seasonDetails.has("episodes")) {
            return 0;
        }

        int count = 0;
        for (JsonNode episodeNode : seasonDetails.get("episodes")) {
            Episode episode = Episode.builder()
                    .numero(episodeNode.get("episode_number").asInt())
                    .titulo(textOrNull(episodeNode, "name"))
                    .duracaoMinutos(episodeNode.has("runtime") && !episodeNode.get("runtime").isNull()
                            ? episodeNode.get("runtime").asInt()
                            : null)
                    .temporada(season)
                    .build();
            episodeRepository.save(episode);
            count++;
        }
        return count;
    }

    private String textOrNull(JsonNode node, String field) {
        if (!node.has(field) || node.get(field).isNull()) return null;
        String value = node.get(field).asText();
        return value.isBlank() ? null : value;
    }
}

File: backend/src/tmdb/tmdb.service.ts
import { Injectable, BadRequestException } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import axios, { AxiosInstance } from 'axios';
import { ContentType } from '@prisma/client';
import { PrismaService } from '../prisma/prisma.service';

interface TmdbSearchResult {
  tmdbId: number;
  titulo: string;
  ano: number | null;
  imagemUrl: string | null;
  sinopse: string | null;
  tipo: ContentType;
  avaliacao: number | null;
}

interface TmdbImportResult {
  conteudoId: number;
  titulo: string;
  temporadasImportadas: number;
  episodiosImportadas: number;
}

/**
 * Cacheia buscas repetidas (mesmo termo + tipo) em memória, evitando bater na
 * API do TMDB toda vez que o usuário reabre a mesma pesquisa.
 * Equivalente ao @Cacheable(value = "tmdbBusca", ...) do backend Java.
 */
class BuscaCache {
  private readonly store = new Map<string, { valor: TmdbSearchResult[]; expiraEm: number }>();
  private readonly ttlMs = 10 * 60 * 1000; // 10 minutos

  get(chave: string): TmdbSearchResult[] | undefined {
    const item = this.store.get(chave);
    if (!item) return undefined;
    if (Date.now() > item.expiraEm) {
      this.store.delete(chave);
      return undefined;
    }
    return item.valor;
  }

  set(chave: string, valor: TmdbSearchResult[]) {
    this.store.set(chave, { valor, expiraEm: Date.now() + this.ttlMs });
  }
}

@Injectable()
export class TmdbService {
  private readonly http: AxiosInstance;
  private readonly apiKey: string;
  private readonly baseUrl: string;
  private readonly imageBaseUrl: string;
  private readonly cache = new BuscaCache();

  constructor(
    private readonly configService: ConfigService,
    private readonly prisma: PrismaService,
  ) {
    this.apiKey = this.configService.get<string>('tmdb.apiKey');
    this.baseUrl = this.configService.get<string>('tmdb.baseUrl');
    this.imageBaseUrl = this.configService.get<string>('tmdb.imageBaseUrl');
    this.http = axios.create({ timeout: 8000 });
  }

  isConfigured(): boolean {
    return !!this.apiKey && this.apiKey.trim() !== '';
  }

  async buscar(query: string, tipo: ContentType): Promise<TmdbSearchResult[]> {
    if (!this.isConfigured()) {
      throw new BadRequestException(
        'Integração com o TMDB não configurada. Defina a variável de ambiente TMDB_API_KEY.',
      );
    }

    const chaveCache = `${tipo}-${query.toLowerCase()}`;
    const emCache = this.cache.get(chaveCache);
    if (emCache) return emCache;

    const path = tipo === ContentType.FILME ? '/search/movie' : '/search/tv';
    const { data } = await this.http.get(`${this.baseUrl}${path}`, {
      params: {
        api_key: this.apiKey,
        language: 'pt-BR',
        query,
        include_adult: false,
      },
    });

    const resultados: TmdbSearchResult[] = (data?.results ?? [])
      .slice(0, 20)
      .map((node: any) => this.toSearchResult(node, tipo));

    this.cache.set(chaveCache, resultados);
    return resultados;
  }

  private toSearchResult(node: any, tipo: ContentType): TmdbSearchResult {
    const titulo = tipo === ContentType.FILME ? node.title : node.name;
    const dataLancamento = tipo === ContentType.FILME ? node.release_date : node.first_air_date;

    let ano: number | null = null;
    if (dataLancamento && dataLancamento.length >= 4) {
      const parsed = parseInt(dataLancamento.substring(0, 4), 10);
      if (!Number.isNaN(parsed)) ano = parsed;
    }

    const imagemUrl = node.poster_path ? this.imageBaseUrl + node.poster_path : null;

    return {
      tmdbId: node.id,
      titulo: titulo || 'Sem título',
      ano,
      imagemUrl,
      sinopse: node.overview || null,
      tipo,
      avaliacao: typeof node.vote_average === 'number' ? node.vote_average : null,
    };
  }

  async importar(tmdbId: number, tipo: ContentType): Promise<TmdbImportResult> {
    if (!this.isConfigured()) {
      throw new BadRequestException(
        'Integração com o TMDB não configurada. Defina a variável de ambiente TMDB_API_KEY.',
      );
    }

    const path = tipo === ContentType.FILME ? `/movie/${tmdbId}` : `/tv/${tmdbId}`;
    let details: any;
    try {
      const response = await this.http.get(`${this.baseUrl}${path}`, {
        params: { api_key: this.apiKey, language: 'pt-BR' },
      });
      details = response.data;
    } catch {
      details = null;
    }
    if (!details) {
      throw new BadRequestException('Não foi possível encontrar este título no TMDB');
    }

    const titulo = tipo === ContentType.FILME ? details.title : details.name;
    const dataLancamento = tipo === ContentType.FILME ? details.release_date : details.first_air_date;
    let ano: number | null = null;
    if (dataLancamento && dataLancamento.length >= 4) {
      const parsed = parseInt(dataLancamento.substring(0, 4), 10);
      if (!Number.isNaN(parsed)) ano = parsed;
    }

    let genero: string | null = null;
    if (Array.isArray(details.genres)) {
      genero = details.genres.map((g: any) => g.name).join(', ');
    }

    const imagemUrl = details.poster_path ? this.imageBaseUrl + details.poster_path : null;

    const content = await this.prisma.content.create({
      data: {
        titulo: titulo || 'Sem título',
        sinopse: details.overview || null,
        genero,
        ano,
        imagemUrl,
        tipo,
        assistido: false,
      },
    });

    let temporadasImportadas = 0;
    let episodiosImportadas = 0;

    if (tipo === ContentType.SERIE && Array.isArray(details.seasons)) {
      for (const seasonNode of details.seasons) {
        const seasonNumber = seasonNode.season_number;
        const season = await this.prisma.season.create({
          data: { numero: seasonNumber, titulo: seasonNode.name || null, conteudoId: content.id },
        });
        temporadasImportadas++;
        episodiosImportadas += await this.importarEpisodios(tmdbId, seasonNumber, season.id);
      }
    }

    return {
      conteudoId: content.id,
      titulo: content.titulo,
      temporadasImportadas,
      episodiosImportadas,
    };
  }

  private async importarEpisodios(tmdbId: number, seasonNumber: number, seasonId: number): Promise<number> {
    let seasonDetails: any;
    try {
      const response = await this.http.get(`${this.baseUrl}/tv/${tmdbId}/season/${seasonNumber}`, {
        params: { api_key: this.apiKey, language: 'pt-BR' },
      });
      seasonDetails = response.data;
    } catch {
      return 0;
    }

    if (!seasonDetails || !Array.isArray(seasonDetails.episodes)) {
      return 0;
    }

    let count = 0;
    for (const episodeNode of seasonDetails.episodes) {
      await this.prisma.episode.create({
        data: {
          numero: episodeNode.episode_number,
          titulo: episodeNode.name || null,
          duracaoMinutos: typeof episodeNode.runtime === 'number' ? episodeNode.runtime : null,
          temporadaId: seasonId,
        },
      });
      count++;
    }
    return count;
  }
}
